Add _clone_dim_order portable kernel

keyprocedure opened the pull request:
### Summary
This is PR 1 of 3 implementing a dim order aware clone op. 

Currently, clone ops are removed during export as no-ops, causing memory layout (dim order) changes to be lost. This can cause backend failures, incorrect outputs when ops expect specific layouts, and performance degradation. This set of PRs introduces a dim order aware clone op, `_clone_dim_order`, which preserves memory layout changes by explicitly storing dim order information. This is implemented by replacing standard clone ops with this variant during export and updating the clone removal transform to preserve clones that change layout.

This PR adds the portable CPU kernel for the `_clone_dim_order` op, implementing a clone variant that preserves dim order at runtime. The portable kernel validates dtype and layout compatibility, resizes the output tensor if needed, and performs an element wise clone of the tensors.

Note: A future PR will add the ATen kernel for `_clone_dim_order`.

[Resource for Dim Order in ExecuTorch](https://github.com/pytorch/executorch/issues/8037)

Related PRs:
- PR 2: [#13735](https://github.com/pytorch/executorch/pull/13735) - Register `_clone_dim_order` op and map `aten.clone`
- PR 3: [#12976](https://github.com/pytorch/executorch/pull/12976) -  Update RemoveCloneOpsTransform to be dim_order aware

Fixes #12645 

### Test plan
Added kernel runtime tests to verify:
- Tensors of all real dtypes are cloned correctly.
- Failure when input and output tensor shapes mismatch.
- Failure with unsupported memory formats.
- Failure when `non_blocking=true` since the portable kernel only supports blocking data transfer.
- Dynamic shape outputs are cloned with correct values.
- Layout conversions are cloned correctly for `contiguous` to `channels_last`, `channels_last` to `contiguous`, and `channels_last` is preserved.

All runtime tests pass via:
`build-ninja/kernels/test/portable_kernels_test`

keyprocedure committed:
Add clone_dim_order kernel (portable + ATen) and layout conversion tests

keyprocedure committed:
Remove duplicate helper functions (moved to utils)

keyprocedure committed:
Revert ATen _clone_dim_order implementation and refactoring

keyprocedure committed:
Register _clone_dim_order and add operator tests

Gasoonjia reviewed (commented):
@keyprocedure 

Good start! Thanks for your great work!

Beyond than what we have now, it would be awesome if we can have a runtime test for the new runtime operator. https://github.com/pytorch/executorch/blob/main/kernels/test/op__to_dim_order_copy_test.cpp is the test for to_dim_order_copy and you can use that as an example.

Also please mark this PR as Open instead of Draft if it is ready to review. Thx!

keyprocedure commented on exir/passes/dim_order_ops_registry.py:
```diff
@@ -28,6 +28,14 @@
     "_empty_dim_order.out(int[] size, *, int[]? dim_order=None, Tensor(a!) out) -> Tensor(a!)"
 )
 
+lib.define(
```
That makes sense, but I needed to register the operator here otherwise the tests I added fail since there is no Python side reference to `_clone_dim_order`

Gasoonjia replied:
This PR should only focus on our runtime changes, so no python side will refer to our new operator. 

> should all the tests for this PR have been on the kernel side and not in [test_memory_format_ops_pass.py](https://github.com/pytorch/executorch/pull/12974/files#diff-06df0587c95ca44af02799ffd147b40e324ce68def041f25e14f98744294db9cR94)

Yes absolutely correct! This PR is only for portable kernel and its tests. Sorry for any misleading!

keyprocedure replied:
I’ve added the runtime test and all local tests passed. I couldn’t run the `DynamicShapeUnbound` test since it depends on `SupportedFeatures` and `supported_features.h` doesn’t seem to be generated in OSS builds.

keyprocedure replied:
Please disregard my previous comment about the missing `SupportedFeatures` dependency, the issue was with my local build setup. All tests pass now.

keyprocedure committed:
Add broadcast_util dep for copy_ops_util.h

keyprocedure committed:
Remove Python side registration and tests for _clone_dim_order

keyprocedure committed:
Add _clone_dim_order runtime test

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-kernel

Gasoonjia commented on kernels/test/op__clone_dim_order_test.cpp:
```diff
@@ -0,0 +1,379 @@
+    }
+  }
+
+  /* %python
```
Please remove unrelated comments with tests

Gasoonjia reviewed (commented):
LGTM! Some minor feedback but majority looks good!

keyprocedure committed:
Clarify dim_order format and remove unrelated comments

keyprocedure commented:
@Gasoonjia CI failed because of a missing dependency in [copy_ops_util.h](https://github.com/pytorch/executorch/pull/12974/files#diff-f91ccb7a0924492d2368a528c7629c3c8dba4d208881846b991f6af5e614db04R12):
`fatal error: 'executorch/kernels/portable/cpu/util/broadcast_util.h' file not found`.

This happened after I refactored a function into copy_ops_util.h and added the import. 
I added the import to [targets.bzl](https://github.com/pytorch/executorch/pull/12974/files#diff-f91ccb7a0924492d2368a528c7629c3c8dba4d208881846b991f6af5e614db04R12), but CI still fails. Do you have insights into what I could be missing?

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-kernel

keyprocedure committed:
Move broadcast_util to exported_deps for copy_ops_util

Gasoonjia commented:
@keyprocedure  so glad you've fixed the issue! Sry for late review. Restart ci.

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-kernel

keyprocedure committed:
Add op__clone_dim_order to op_library

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-kernel

keyprocedure commented:
> @keyprocedure so glad you've fixed the issue! Sry for late review. Restart ci.

No worries, I appreciate all the support :)

Progress with CI: 
The 10 failing tests were all due to a link error:
`Action failed: root//examples/portable/executor_runner:executor_runner (cxx_link_executable)`
I've registered `_clone_dim_order` in [op_registration_util.bzl](https://github.com/pytorch/executorch/pull/12974/files#diff-20ab8922e8d76d51917a0a03646ae7aa25e958bc2a2c9d0e26f438a9e483d814R1332) and I can successfully build executor_runner locally. 

Do you have any recommendations on which targets I should build locally to ensure everything that relies on new ops will build successfully or is there a Docker image available to run CI locally? I've tried to build entire dirs such as `//examples/portable/...` but run into dep failures unrelated to this PR.

Gasoonjia commented on shim_et/xplat/executorch/kernels/portable/op_registration_util.bzl:
```diff
@@ -1329,6 +1329,13 @@ ATEN_OPS = (
             "//executorch/kernels/portable/cpu/util:copy_ops_util",
```
we can remove broadcast_util here right? Since copy_util will depend on it

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-kernel

Gasoonjia commented:
thansk for your great work! @keyprocedure 
I don't think there's one single point target you can built for running CI locally. Thanks for bring it up! I think it is a good suggestion we can work on in the future for better local ci coverage!

keyprocedure commented:
> BTW Have you tried this [`main`/CONTRIBUTING.md#testing](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md?rgh-link-date=2025-08-08T23%3A07%3A19Z#testing) ? @keyprocedure

Thanks for sharing this!

After stopping some warnings from causing the build to fail, I ran the test script and everything passes except the [PyTree EmptySpec test](https://github.com/pytorch/executorch/blob/50ffe4466b1fb6d737c7f1359c4db08c4f734ad3/extension/pytree/test/test_pytree.cpp#L202), which seems unrelated. 

But I'll run this script to validate future changes.

Gasoonjia committed:
Merge branch 'main' into add-dim-order-clone-kernel
